{"expand":"renderedFields,names,schema,transitions,operations,editmeta,changelog","id":"12504344","self":"https://issues.apache.org/jira/rest/api/latest/issue/12504344","key":"DERBY-5192","fields":{"issuetype":{"self":"https://issues.apache.org/jira/rest/api/2/issuetype/1","id":"1","description":"A problem which impairs or prevents the functions of the product.","iconUrl":"https://issues.apache.org/jira/images/icons/issuetypes/bug.png","name":"Bug","subtask":false},"timespent":null,"project":{"self":"https://issues.apache.org/jira/rest/api/2/project/10594","id":"10594","key":"DERBY","name":"Derby","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/projectavatar?pid=10594&avatarId=10122","24x24":"https://issues.apache.org/jira/secure/projectavatar?size=small&pid=10594&avatarId=10122","16x16":"https://issues.apache.org/jira/secure/projectavatar?size=xsmall&pid=10594&avatarId=10122","32x32":"https://issues.apache.org/jira/secure/projectavatar?size=medium&pid=10594&avatarId=10122"},"projectCategory":{"self":"https://issues.apache.org/jira/rest/api/2/projectCategory/10090","id":"10090","description":"DB related projects","name":"DB"}},"fixVersions":[{"self":"https://issues.apache.org/jira/rest/api/2/version/12316362","id":"12316362","description":"First release on the 10.8 branch","name":"10.8.1.2","archived":false,"released":true,"releaseDate":"2011-04-29"},{"self":"https://issues.apache.org/jira/rest/api/2/version/12316344","id":"12316344","description":"First release on the 10.9 branch","name":"10.9.1.0","archived":false,"released":true,"releaseDate":"2012-06-25"}],"aggregatetimespent":null,"resolution":{"self":"https://issues.apache.org/jira/rest/api/2/resolution/1","id":"1","description":"A fix for this issue is checked into the tree and tested.","name":"Fixed"},"customfield_12310220":"2011-04-15 13:05:49.221","customfield_12312322":null,"customfield_12312323":null,"customfield_12310420":"24705","customfield_12310222":"3_*:*_1_*:*_249873432_*|*_1_*:*_1_*:*_7561242_*|*_5_*:*_1_*:*_0","customfield_12312320":null,"customfield_12312321":null,"customfield_12312120":null,"customfield_12312121":null,"resolutiondate":"2011-04-18T07:42:12.576+0000","workratio":-1,"customfield_12312328":null,"customfield_12312329":null,"customfield_12312326":null,"customfield_12310300":null,"customfield_12312327":null,"customfield_12312324":null,"customfield_12312325":null,"lastViewed":null,"watches":{"self":"https://issues.apache.org/jira/rest/api/2/issue/DERBY-5192/watchers","watchCount":0,"isWatching":false},"created":"2011-04-15T08:11:38.021+0000","priority":{"self":"https://issues.apache.org/jira/rest/api/2/priority/3","iconUrl":"https://issues.apache.org/jira/images/icons/priorities/major.png","name":"Major","id":"3"},"labels":[],"customfield_12312333":null,"customfield_12312334":null,"customfield_12310310":"1.0","customfield_12312331":null,"customfield_12312332":null,"timeestimate":null,"aggregatetimeoriginalestimate":null,"versions":[{"self":"https://issues.apache.org/jira/rest/api/2/version/12316362","id":"12316362","description":"First release on the 10.8 branch","name":"10.8.1.2","archived":false,"released":true,"releaseDate":"2011-04-29"}],"customfield_12311120":null,"customfield_12312330":null,"issuelinks":[{"id":"12338588","self":"https://issues.apache.org/jira/rest/api/2/issueLink/12338588","type":{"id":"10030","name":"Reference","inward":"is related to","outward":"relates to","self":"https://issues.apache.org/jira/rest/api/2/issueLinkType/10030"},"outwardIssue":{"id":"12425869","key":"DERBY-4238","self":"https://issues.apache.org/jira/rest/api/2/issue/12425869","fields":{"summary":"JMX  test failure starting network server doesn't  seem to get recorded to the fail directory ","status":{"self":"https://issues.apache.org/jira/rest/api/2/status/1","description":"The issue is open and ready for the assignee to start work on it.","iconUrl":"https://issues.apache.org/jira/images/icons/statuses/open.png","name":"Open","id":"1","statusCategory":{"self":"https://issues.apache.org/jira/rest/api/2/statuscategory/2","id":2,"key":"new","colorName":"blue-gray","name":"New"}},"priority":{"self":"https://issues.apache.org/jira/rest/api/2/priority/4","iconUrl":"https://issues.apache.org/jira/images/icons/priorities/minor.png","name":"Minor","id":"4"},"issuetype":{"self":"https://issues.apache.org/jira/rest/api/2/issuetype/1","id":"1","description":"A problem which impairs or prevents the functions of the product.","iconUrl":"https://issues.apache.org/jira/images/icons/issuetypes/bug.png","name":"Bug","subtask":false}}}},{"id":"12338591","self":"https://issues.apache.org/jira/rest/api/2/issueLink/12338591","type":{"id":"10030","name":"Reference","inward":"is related to","outward":"relates to","self":"https://issues.apache.org/jira/rest/api/2/issueLinkType/10030"},"outwardIssue":{"id":"12468858","key":"DERBY-4739","self":"https://issues.apache.org/jira/rest/api/2/issue/12468858","fields":{"summary":"test hang in Suites.all starting derbynet.SSLTest on Mac OS X 10.6","status":{"self":"https://issues.apache.org/jira/rest/api/2/status/5","description":"A resolution has been taken, and it is awaiting verification by reporter. From here issues are either reopened, or are closed.","iconUrl":"https://issues.apache.org/jira/images/icons/statuses/resolved.png","name":"Resolved","id":"5","statusCategory":{"self":"https://issues.apache.org/jira/rest/api/2/statuscategory/3","id":3,"key":"done","colorName":"green","name":"Complete"}},"priority":{"self":"https://issues.apache.org/jira/rest/api/2/priority/3","iconUrl":"https://issues.apache.org/jira/images/icons/priorities/major.png","name":"Major","id":"3"},"issuetype":{"self":"https://issues.apache.org/jira/rest/api/2/issuetype/1","id":"1","description":"A problem which impairs or prevents the functions of the product.","iconUrl":"https://issues.apache.org/jira/images/icons/issuetypes/bug.png","name":"Bug","subtask":false}}}},{"id":"12338586","self":"https://issues.apache.org/jira/rest/api/2/issueLink/12338586","type":{"id":"10030","name":"Reference","inward":"is related to","outward":"relates to","self":"https://issues.apache.org/jira/rest/api/2/issueLinkType/10030"},"outwardIssue":{"id":"12423430","key":"DERBY-4176","self":"https://issues.apache.org/jira/rest/api/2/issue/12423430","fields":{"summary":"java.net.SocketException: Connection reset with ibm 1.5. running ServerPropertiesTest in suites.All","status":{"self":"https://issues.apache.org/jira/rest/api/2/status/6","description":"The issue is considered finished, the resolution is correct. Issues which are closed can be reopened.","iconUrl":"https://issues.apache.org/jira/images/icons/statuses/closed.png","name":"Closed","id":"6","statusCategory":{"self":"https://issues.apache.org/jira/rest/api/2/statuscategory/3","id":3,"key":"done","colorName":"green","name":"Complete"}},"priority":{"self":"https://issues.apache.org/jira/rest/api/2/priority/3","iconUrl":"https://issues.apache.org/jira/images/icons/priorities/major.png","name":"Major","id":"3"},"issuetype":{"self":"https://issues.apache.org/jira/rest/api/2/issuetype/1","id":"1","description":"A problem which impairs or prevents the functions of the product.","iconUrl":"https://issues.apache.org/jira/images/icons/issuetypes/bug.png","name":"Bug","subtask":false}}}},{"id":"12338589","self":"https://issues.apache.org/jira/rest/api/2/issueLink/12338589","type":{"id":"10030","name":"Reference","inward":"is related to","outward":"relates to","self":"https://issues.apache.org/jira/rest/api/2/issueLinkType/10030"},"outwardIssue":{"id":"12427599","key":"DERBY-4265","self":"https://issues.apache.org/jira/rest/api/2/issue/12427599","fields":{"summary":"timed out waiting for server to start in  StatementPoolingTest","status":{"self":"https://issues.apache.org/jira/rest/api/2/status/6","description":"The issue is considered finished, the resolution is correct. Issues which are closed can be reopened.","iconUrl":"https://issues.apache.org/jira/images/icons/statuses/closed.png","name":"Closed","id":"6","statusCategory":{"self":"https://issues.apache.org/jira/rest/api/2/statuscategory/3","id":3,"key":"done","colorName":"green","name":"Complete"}},"priority":{"self":"https://issues.apache.org/jira/rest/api/2/priority/3","iconUrl":"https://issues.apache.org/jira/images/icons/priorities/major.png","name":"Major","id":"3"},"issuetype":{"self":"https://issues.apache.org/jira/rest/api/2/issuetype/1","id":"1","description":"A problem which impairs or prevents the functions of the product.","iconUrl":"https://issues.apache.org/jira/images/icons/issuetypes/bug.png","name":"Bug","subtask":false}}}},{"id":"12338590","self":"https://issues.apache.org/jira/rest/api/2/issueLink/12338590","type":{"id":"10030","name":"Reference","inward":"is related to","outward":"relates to","self":"https://issues.apache.org/jira/rest/api/2/issueLinkType/10030"},"outwardIssue":{"id":"12430960","key":"DERBY-4319","self":"https://issues.apache.org/jira/rest/api/2/issue/12430960","fields":{"summary":"hang in suites.all with ibm 1.5 on AIX after ttestDefaultProperties","status":{"self":"https://issues.apache.org/jira/rest/api/2/status/6","description":"The issue is considered finished, the resolution is correct. Issues which are closed can be reopened.","iconUrl":"https://issues.apache.org/jira/images/icons/statuses/closed.png","name":"Closed","id":"6","statusCategory":{"self":"https://issues.apache.org/jira/rest/api/2/statuscategory/3","id":3,"key":"done","colorName":"green","name":"Complete"}},"priority":{"self":"https://issues.apache.org/jira/rest/api/2/priority/3","iconUrl":"https://issues.apache.org/jira/images/icons/priorities/major.png","name":"Major","id":"3"},"issuetype":{"self":"https://issues.apache.org/jira/rest/api/2/issuetype/1","id":"1","description":"A problem which impairs or prevents the functions of the product.","iconUrl":"https://issues.apache.org/jira/images/icons/issuetypes/bug.png","name":"Bug","subtask":false}}}},{"id":"12338593","self":"https://issues.apache.org/jira/rest/api/2/issueLink/12338593","type":{"id":"10030","name":"Reference","inward":"is related to","outward":"relates to","self":"https://issues.apache.org/jira/rest/api/2/issueLinkType/10030"},"outwardIssue":{"id":"12504367","key":"DERBY-5194","self":"https://issues.apache.org/jira/rest/api/2/issue/12504367","fields":{"summary":"AutoloadTest timed out waiting for network server to start","status":{"self":"https://issues.apache.org/jira/rest/api/2/status/6","description":"The issue is considered finished, the resolution is correct. Issues which are closed can be reopened.","iconUrl":"https://issues.apache.org/jira/images/icons/statuses/closed.png","name":"Closed","id":"6","statusCategory":{"self":"https://issues.apache.org/jira/rest/api/2/statuscategory/3","id":3,"key":"done","colorName":"green","name":"Complete"}},"priority":{"self":"https://issues.apache.org/jira/rest/api/2/priority/3","iconUrl":"https://issues.apache.org/jira/images/icons/priorities/major.png","name":"Major","id":"3"},"issuetype":{"self":"https://issues.apache.org/jira/rest/api/2/issuetype/1","id":"1","description":"A problem which impairs or prevents the functions of the product.","iconUrl":"https://issues.apache.org/jira/images/icons/issuetypes/bug.png","name":"Bug","subtask":false}}}},{"id":"12338587","self":"https://issues.apache.org/jira/rest/api/2/issueLink/12338587","type":{"id":"10030","name":"Reference","inward":"is related to","outward":"relates to","self":"https://issues.apache.org/jira/rest/api/2/issueLinkType/10030"},"outwardIssue":{"id":"12424187","key":"DERBY-4201","self":"https://issues.apache.org/jira/rest/api/2/issue/12424187","fields":{"summary":"SecureServerTest AssertionFailedError: Timed out waiting for network server to start","status":{"self":"https://issues.apache.org/jira/rest/api/2/status/6","description":"The issue is considered finished, the resolution is correct. Issues which are closed can be reopened.","iconUrl":"https://issues.apache.org/jira/images/icons/statuses/closed.png","name":"Closed","id":"6","statusCategory":{"self":"https://issues.apache.org/jira/rest/api/2/statuscategory/3","id":3,"key":"done","colorName":"green","name":"Complete"}},"priority":{"self":"https://issues.apache.org/jira/rest/api/2/priority/4","iconUrl":"https://issues.apache.org/jira/images/icons/priorities/minor.png","name":"Minor","id":"4"},"issuetype":{"self":"https://issues.apache.org/jira/rest/api/2/issuetype/1","id":"1","description":"A problem which impairs or prevents the functions of the product.","iconUrl":"https://issues.apache.org/jira/images/icons/issuetypes/bug.png","name":"Bug","subtask":false}}}},{"id":"12338592","self":"https://issues.apache.org/jira/rest/api/2/issueLink/12338592","type":{"id":"10030","name":"Reference","inward":"is related to","outward":"relates to","self":"https://issues.apache.org/jira/rest/api/2/issueLinkType/10030"},"outwardIssue":{"id":"12480807","key":"DERBY-4914","self":"https://issues.apache.org/jira/rest/api/2/issue/12480807","fields":{"summary":"Utf8CcsidManagerClientTest timed out waiting for network server to start","status":{"self":"https://issues.apache.org/jira/rest/api/2/status/6","description":"The issue is considered finished, the resolution is correct. Issues which are closed can be reopened.","iconUrl":"https://issues.apache.org/jira/images/icons/statuses/closed.png","name":"Closed","id":"6","statusCategory":{"self":"https://issues.apache.org/jira/rest/api/2/statuscategory/3","id":3,"key":"done","colorName":"green","name":"Complete"}},"priority":{"self":"https://issues.apache.org/jira/rest/api/2/priority/4","iconUrl":"https://issues.apache.org/jira/images/icons/priorities/minor.png","name":"Minor","id":"4"},"issuetype":{"self":"https://issues.apache.org/jira/rest/api/2/issuetype/1","id":"1","description":"A problem which impairs or prevents the functions of the product.","iconUrl":"https://issues.apache.org/jira/images/icons/issuetypes/bug.png","name":"Bug","subtask":false}}}}],"customfield_12312339":null,"assignee":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=knutanders","name":"knutanders","emailAddress":"knut dot hatlen at oracle dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Knut Anders Hatlen","active":true},"customfield_12312337":null,"customfield_12312338":null,"updated":"2011-06-30T09:21:40.302+0000","customfield_12312335":null,"customfield_12311720":null,"customfield_12312336":null,"status":{"self":"https://issues.apache.org/jira/rest/api/2/status/6","description":"The issue is considered finished, the resolution is correct. Issues which are closed can be reopened.","iconUrl":"https://issues.apache.org/jira/images/icons/statuses/closed.png","name":"Closed","id":"6","statusCategory":{"self":"https://issues.apache.org/jira/rest/api/2/statuscategory/3","id":3,"key":"done","colorName":"green","name":"Complete"}},"components":[{"self":"https://issues.apache.org/jira/rest/api/2/component/11410","id":"11410","name":"Network Server"}],"timeoriginalestimate":null,"description":"I've seen on three occurrences with the 10.8.1.0 and 10.8.1.1 release candidates that suites.All has been stuck when setting up the network server decorator for the management test suite. Here's what I see at the end of the console output (running with derby.tests.trace=true):\r\n\r\n    [junit] test_jdbc4_1 used 0 ms \r\n    [junit] test_jdbc4_1 used 0 ms \r\n    [junit] test_notBooted used 1623 ms java.net.SocketException: Connection reset\r\n    [junit] \tat java.net.SocketInputStream.read(SocketInputStream.java:189)\r\n    [junit] \tat java.net.SocketInputStream.read(SocketInputStream.java:121)\r\n    [junit] \tat java.net.SocketInputStream.read(SocketInputStream.java:107)\r\n    [junit] \tat org.apache.derby.impl.drda.NetworkServerControlImpl.fillReplyBuffer(Unknown Source)\r\n    [junit] \tat org.apache.derby.impl.drda.NetworkServerControlImpl.readResult(Unknown Source)\r\n    [junit] \tat org.apache.derby.impl.drda.NetworkServerControlImpl.pingWithNoOpen(Unknown Source)\r\n    [junit] \tat org.apache.derby.impl.drda.NetworkServerControlImpl.ping(Unknown Source)\r\n    [junit] \tat org.apache.derby.drda.NetworkServerControl.ping(Unknown Source)\r\n    [junit] \tat org.apache.derbyTesting.junit.NetworkServerTestSetup.pingForServerUp(NetworkServerTestSetup.java:567)\r\n    [junit] \tat org.apache.derbyTesting.junit.NetworkServerTestSetup.pingForServerStart(NetworkServerTestSetup.java:636)\r\n    [junit] \tat org.apache.derbyTesting.junit.NetworkServerTestSetup.setUp(NetworkServerTestSetup.java:196)\r\n    [junit] \tat junit.extensions.TestSetup$1.protect(TestSetup.java:20)\r\n    [junit] \tat junit.framework.TestResult.runProtected(TestResult.java:124)\r\n    [junit] \tat junit.extensions.TestSetup.run(TestSetup.java:25)\r\n    [junit] \tat org.apache.derbyTesting.junit.BaseTestSetup.run(BaseTestSetup.java:57)\r\n    [junit] \tat junit.extensions.TestDecorator.basicRun(TestDecorator.java:24)\r\n    [junit] \tat junit.extensions.TestSetup$1.protect(TestSetup.java:21)\r\n    [junit] \tat junit.framework.TestResult.runProtected(TestResult.java:124)\r\n    [junit] \tat junit.extensions.TestSetup.run(TestSetup.java:25)\r\n    [junit] \tat junit.extensions.TestDecorator.basicRun(TestDecorator.java:24)\r\n    [junit] \tat junit.extensions.TestSetup$1.protect(TestSetup.java:21)\r\n    [junit] \tat junit.framework.TestResult.runProtected(TestResult.java:124)\r\n    [junit] \tat junit.extensions.TestSetup.run(TestSetup.java:25)\r\n    [junit] \tat org.apache.derbyTesting.junit.BaseTestSetup.run(BaseTestSetup.java:57)\r\n    [junit] \tat junit.framework.TestSuite.runTest(TestSuite.java:230)\r\n    [junit] \tat junit.framework.TestSuite.run(TestSuite.java:225)\r\n    [junit] \tat junit.framework.TestSuite.runTest(TestSuite.java:230)\r\n    [junit] \tat junit.framework.TestSuite.run(TestSuite.java:225)\r\n    [junit] \tat junit.framework.TestSuite.runTest(TestSuite.java:230)\r\n    [junit] \tat junit.framework.TestSuite.run(TestSuite.java:225)\r\n    [junit] \tat junit.framework.TestSuite.runTest(TestSuite.java:230)\r\n    [junit] \tat junit.framework.TestSuite.run(TestSuite.java:225)\r\n    [junit] \tat org.apache.tools.ant.taskdefs.optional.junit.JUnitTestRunner.run(JUnitTestRunner.java:518)\r\n    [junit] \tat org.apache.tools.ant.taskdefs.optional.junit.JUnitTestRunner.launch(JUnitTestRunner.java:1052)\r\n    [junit] \tat org.apache.tools.ant.taskdefs.optional.junit.JUnitTestRunner.main(JUnitTestRunner.java:906)\r\n\r\nThe exception trace is just printed to the console, but it doesn't make any test fail. (The exception printed on FreeBSD was different, it said \"DRDA_InvalidReplyTooShort.S:Invalid reply from network server: Insufficient data.\" The other two looked like the one above.)\r\n\r\nAll the hangs have happened on VirtualBox instances, though with different guest operating systems and JVMs. Probably the timing is different from what we have on physical machines.","customfield_10010":null,"timetracking":{},"customfield_12312026":null,"customfield_12310200":null,"customfield_12312023":null,"customfield_12312024":null,"customfield_12312340":null,"aggregatetimeestimate":null,"customfield_12312022":null,"customfield_12312341":null,"customfield_12312220":null,"customfield_12310921":null,"customfield_12310920":"36397","summary":"Setting up network server for management tests hangs intermittently","creator":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=knutanders","name":"knutanders","emailAddress":"knut dot hatlen at oracle dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Knut Anders Hatlen","active":true},"subtasks":[],"customfield_12310090":null,"reporter":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=knutanders","name":"knutanders","emailAddress":"knut dot hatlen at oracle dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Knut Anders Hatlen","active":true},"customfield_12310291":null,"customfield_12310290":null,"aggregateprogress":{"progress":0,"total":0},"environment":"Derby 10.8.1.0 - FreeBSD 8.2 (i386) - OpenJDK 6 (b20)\r\nDerby 10.8.1.0 - Oracle Enterprise Linux 6.0 (x86_64) - OpenJDK 6 (b17)\r\nDerby 10.8.1.1 - Debian GNU/Linux 6.0.1 (i386) - JDK 7 (build 1.7.0-ea-b135)","customfield_12311020":null,"customfield_12310050":null,"duedate":null,"customfield_12310250":null,"progress":{"progress":0,"total":0},"comment":{"startAt":0,"maxResults":7,"total":7,"comments":[{"self":"https://issues.apache.org/jira/rest/api/2/issue/12504344/comment/13020253","id":"13020253","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=knutanders","name":"knutanders","emailAddress":"knut dot hatlen at oracle dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Knut Anders Hatlen","active":true},"body":"When looking at the thread dump for the hung process, I see the\r\nfollowing:\r\n\r\n- There is a server thread that's blocked on a call to\r\n  shutdownSync.wait() in NetworkServerControlImpl.blockingStart().\r\n  This means the network server is up and running, waiting for a\r\n  shutdown signal.\r\n\r\n- There is no ClientThread, whereas one would expect one if the\r\n  network server was up. This is the thread that accepts incoming\r\n  connections from the clients.\r\n\r\n- The main JUnit thread is stuck in a read call on the client socket.\r\n  This is consistent with what one would expect when there is no\r\n  ClientThread. The client is able to connect to the server, but the\r\n  server won't process the connection.\r\n\r\nInspecting a heap dump of the process, I notice that the shutdown flag\r\nin the NetworkServerControlImpl instance is actually true. So it looks\r\nlike the server has been shut down, but it's still waiting for the\r\nshutdown signal. The only way I can see that this can happen, is if\r\nthe shutdown signal is sent before blockingStart() has come to the\r\nshutdownSync.wait() call. The wait/notify code here isn't quite\r\nidiomatic. The wait() call isn't placed inside a while loop checking\r\nthe wait condition, so if the shutdown happens too early, the server\r\nwon't ever notice the notifyAll() call that tells it to stop waiting.\r\n\r\nWe have many existing bug reports about problems when starting the\r\nnetwork server from NetworkServerTestSetup (hangs, exceptions printed\r\nby ping(), and timeouts waiting for the network server to start). Some\r\nof these may have the same underlying cause as this bug. A quick JIRA\r\nsearch gave these hits: DERBY-4176, DERBY-4201, DERBY-4238,\r\nDERBY-4265, DERBY-4319, DERBY-4739, DERBY-4914.","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=knutanders","name":"knutanders","emailAddress":"knut dot hatlen at oracle dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Knut Anders Hatlen","active":true},"created":"2011-04-15T09:33:46.257+0000","updated":"2011-04-15T09:33:46.257+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12504344/comment/13020260","id":"13020260","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=knutanders","name":"knutanders","emailAddress":"knut dot hatlen at oracle dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Knut Anders Hatlen","active":true},"body":"The attached patch changes the wait/notify code to use the more common pattern of doing the wait inside a loop that checks the wait condition. This should have the following advantages over the the current code:\r\n\r\n1) If the network server is shut down before blockingStart() gets to the wait() call, it won't hang indefinitely.\r\n\r\n2) If a spurious wakeup happen in the wait() call (which may happen, according to the Java SE javadocs) blockingStart() won't return prematurely.","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=knutanders","name":"knutanders","emailAddress":"knut dot hatlen at oracle dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Knut Anders Hatlen","active":true},"created":"2011-04-15T10:15:33.099+0000","updated":"2011-04-15T10:15:33.099+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12504344/comment/13020264","id":"13020264","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=knutanders","name":"knutanders","emailAddress":"knut dot hatlen at oracle dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Knut Anders Hatlen","active":true},"body":"I tested the patch manually the following way:\r\n\r\n1) In blockingStart(), I added a 20 second sleep right before the synchronized block around the wait() call.\r\n\r\n2) Started a network server using the CLI.\r\n\r\n3) While blockingStart() was in the 20 second sleep, shut down the network server using the CLI.\r\n\r\n4) Ping the network server using the CLI.\r\n\r\nWhat I observed was:\r\n\r\nI) Without the patch: The shutdown command in (3) would complete successfully. However, the network server process would continue running. The first ping in (4) would fail with a \"connection reset\" error. Subsequent pings would hang indefinitely.\r\n\r\nII) With the patch: The shutdown command in (3) completed successfully, and the network server process would terminate once the 20 second sleep was completed. When pinging the network server after the shutdown, a \"connection reset\" error was seen. Subsequent pings would hang until the 20 second sleep was completed and then fail with a \"connection reset\" error. Pinging the server after the server process had terminated failed with \"connection refused\".\r\n\r\nSo, in short: Without the patch, the server isn't properly shut down, and incoming connection attempts will hang indefinitely. With the patch, the server is shut down, and incoming connection attempts will fail instead of hanging.","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=knutanders","name":"knutanders","emailAddress":"knut dot hatlen at oracle dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Knut Anders Hatlen","active":true},"created":"2011-04-15T10:42:25.491+0000","updated":"2011-04-15T10:42:25.491+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12504344/comment/13020294","id":"13020294","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=dagw","name":"dagw","emailAddress":"dag dot wanvik at oracle dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?ownerId=dagw&avatarId=16789","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&ownerId=dagw&avatarId=16789","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&ownerId=dagw&avatarId=16789","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&ownerId=dagw&avatarId=16789"},"displayName":"Dag H. Wanvik","active":true},"body":"I did look at this one too and agree with you analysis. Patch looks right. +1\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=dagw","name":"dagw","emailAddress":"dag dot wanvik at oracle dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?ownerId=dagw&avatarId=16789","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&ownerId=dagw&avatarId=16789","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&ownerId=dagw&avatarId=16789","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&ownerId=dagw&avatarId=16789"},"displayName":"Dag H. Wanvik","active":true},"created":"2011-04-15T13:05:49.221+0000","updated":"2011-04-15T13:05:49.221+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12504344/comment/13020305","id":"13020305","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"body":"I recall having seen the occasional hang such as this in the past. It never occurred to me to\r\nuse the heap dump to check the status of global variables like that -- neat idea!\r\n\r\nYour change seems good to me, too. If I understand your description, although this\r\nis a generic bug, it's most likely to affect only tests, not \"normal\" usage, because only\r\ntests are likely to try to shut the NS down so soon after starting it, is that right?\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=bryanpendleton","name":"bryanpendleton","emailAddress":"bpendleton dot derby at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Bryan Pendleton","active":true},"created":"2011-04-15T14:02:39.154+0000","updated":"2011-04-15T14:02:39.154+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12504344/comment/13020828","id":"13020828","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=knutanders","name":"knutanders","emailAddress":"knut dot hatlen at oracle dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Knut Anders Hatlen","active":true},"body":"Thanks, Dag and Bryan, for looking at the patch.\r\n\r\nI think Bryan is right that this bug is not likely to affect servers in production, since they usually run for a longer time before they're shut down. The one thing that could also affect production servers, is the lack of protection against spurious wakeups from the wait() call. The javadoc for wait() says that wait() could return even when notify() and notifyAll() haven't been called, but blockingStart() doesn't account for that and treats every wakeup as a shutdown signal.","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=knutanders","name":"knutanders","emailAddress":"knut dot hatlen at oracle dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Knut Anders Hatlen","active":true},"created":"2011-04-17T16:41:21.011+0000","updated":"2011-04-17T16:41:21.011+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12504344/comment/13020927","id":"13020927","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=knutanders","name":"knutanders","emailAddress":"knut dot hatlen at oracle dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Knut Anders Hatlen","active":true},"body":"Committed to trunk with revision 1094315.\r\nCommitted to 10.8 with revision 1094329.","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=knutanders","name":"knutanders","emailAddress":"knut dot hatlen at oracle dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Knut Anders Hatlen","active":true},"created":"2011-04-18T07:42:12.597+0000","updated":"2011-04-18T07:42:12.597+0000"}]},"votes":{"self":"https://issues.apache.org/jira/rest/api/2/issue/DERBY-5192/votes","votes":0,"hasVoted":false},"worklog":{"startAt":0,"maxResults":20,"total":0,"worklogs":[]},"customfield_12311820":"0|i06ldj:"}}